File: src/climradar.py
"""Source ClimRadar — API JSON (/api/stock + /api/live)."""



from __future__ import annotations



import math

import time

from dataclasses import dataclass



import requests



CLIMRADAR_PRODUCT_ID = "portasplit"

CLIMRADAR_STOCK_URL = "https://climradar.fr/api/stock"

CLIMRADAR_LIVE_URL = f"https://climradar.fr/api/live?kind=product&id={CLIMRADAR_PRODUCT_ID}"



RETAILER_MAP: dict[str, str] = {

    "leroy-merlin": "leroy_merlin",

    "castorama": "castorama",

    "boulanger": "boulanger",

    "darty": "darty",

    "fnac": "fnac",

    "manomano": "manomano",

    "amazon": "amazon",

}



DELIVERY_RETAILERS = frozenset(

    {"darty", "boulanger", "castorama", "optimea", "amazon", "fnac", "manomano"}

)



_HEADERS = {

    "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) Chrome/122.0.0.0",

    "Accept": "application/json",

    "Accept-Language": "fr-FR,fr;q=0.9",

}



_stock_cache: dict | None = None

_stock_cache_at: float = 0.0

_STOCK_CACHE_SEC = 60.0





@dataclass

class ClimRadarOffer:

    seller: str

    retailer_id: str

    in_stock: bool

    price: float | None

    url: str

    area: str

    last_update_min: int | None = None





def _fetch_stock_payload(timeout: int = 60, force: bool = False) -> dict | None:

    global _stock_cache, _stock_cache_at

    now = time.monotonic()

    if not force and _stock_cache and now - _stock_cache_at < _STOCK_CACHE_SEC:

        return _stock_cache



    response = requests.get(CLIMRADAR_STOCK_URL, headers=_HEADERS, timeout=timeout)

    response.raise_for_status()

    payload = response.json()

    if not payload.get("stores") or not payload.get("stockByStore"):

        return None



    _stock_cache = payload

    _stock_cache_at = now

    return payload





def fetch_live_rows(timeout: int = 25) -> list[dict]:

    try:

        response = requests.get(CLIMRADAR_LIVE_URL, headers=_HEADERS, timeout=timeout)

        response.raise_for_status()

        return response.json().get("rows", [])

    except requests.RequestException:

        return []





def product_url_map(live_rows: list[dict]) -> dict[str, str]:

    urls: dict[str, str] = {}

    for row in live_rows:

        product_url = row.get("productUrl") or ""

        if not product_url.startswith("http"):

            continue

        store = row.get("store") or {}

        retailer_id = store.get("retailerId") or ""

        store_id = store.get("id") or ""

        if retailer_id:

            urls[retailer_id] = product_url

        if store_id:

            urls[store_id] = product_url

    return urls





def entry_for_product(entries: list[dict]) -> dict | None:

    for entry in entries:

        if entry.get("productId") == CLIMRADAR_PRODUCT_ID:

            return entry

    return entries[0] if entries else None





def _parse_online_offers(payload: dict, live_rows: list[dict]) -> list[ClimRadarOffer]:

    stores_by_id = {s["id"]: s for s in payload.get("stores", [])}

    retailers_by_id = {r["id"]: r for r in payload.get("retailers", [])}

    product_urls = product_url_map(live_rows)

    offers: list[ClimRadarOffer] = []

    seen: set[str] = set()



    for store_id, entries in payload.get("stockByStore", {}).items():

        store = stores_by_id.get(store_id)

        if not store or store.get("channel") != "online" or store.get("country") != "FR":

            continue



        entry = entry_for_product(entries)

        if not entry:

            continue



        slug = store.get("retailerId") or ""

        retailer_id = RETAILER_MAP.get(slug)

        if not retailer_id or retailer_id in seen:

            continue

        seen.add(retailer_id)



        retailer = retailers_by_id.get(slug, {})

        in_stock = entry.get("status") == "en_stock"

        url = product_urls.get(slug) or product_urls.get(store_id) or ""



        offers.append(

            ClimRadarOffer(

                seller=retailer.get("name") or store.get("name", slug),

                retailer_id=retailer_id,

                in_stock=in_stock,

                price=float(entry["price"]) if entry.get("price") is not None else None,

                url=url,

                area="FR",

                last_update_min=entry.get("lastSeenMinAgo"),

            )

        )



    return offers





def is_climradar_online_available(payload: dict | None = None) -> bool:

    if payload is None:

        try:

            payload = _fetch_stock_payload()

        except requests.RequestException:

            return False

    return bool(payload and payload.get("stores") and payload.get("stockByStore"))





def fetch_climradar_offers(timeout: int = 60) -> list[ClimRadarOffer]:

    payload = _fetch_stock_payload(timeout=timeout)

    if not payload:

        return []

    live_rows = fetch_live_rows(timeout=min(timeout, 25))

    return _parse_online_offers(payload, live_rows)





def get_climradar_stock(retailer_id: str) -> ClimRadarOffer | None:

    for offer in fetch_climradar_offers():

        if offer.retailer_id == retailer_id:

            return offer

    return None





def haversine_km(lat1: float, lon1: float, lat2: float, lon2: float) -> float:

    r = 6371.0

    p1, p2 = math.radians(lat1), math.radians(lat2)

    dphi = math.radians(lat2 - lat1)

    dlmb = math.radians(lon2 - lon1)

    a = math.sin(dphi / 2) ** 2 + math.cos(p1) * math.cos(p2) * math.sin(dlmb / 2) ** 2

    return 2 * r * math.asin(math.sqrt(a))





def fetch_climradar_stock_payload(timeout: int = 60, force: bool = False) -> dict | None:

    """Expose le dump /api/stock pour local_stores."""

    try:

        return _fetch_stock_payload(timeout=timeout, force=force)

    except requests.RequestException:

        return None

File: mobile/src/services/climradar.ts
import { productUrlForRetailer } from '../constants/retailers';
import type { LocalStoreOffer, OnlineOffer } from '../types';

export const CLIMRADAR_PRODUCT_ID = 'portasplit';
export const CLIMRADAR_STOCK_URL = 'https://climradar.fr/api/stock';
export const CLIMRADAR_LIVE_URL = `https://climradar.fr/api/live?kind=product&id=${CLIMRADAR_PRODUCT_ID}`;

export const SLUG_TO_RETAILER_ID: Record<string, string> = {
  'leroy-merlin': 'leroy_merlin',
  castorama: 'castorama',
  boulanger: 'boulanger',
  darty: 'darty',
  fnac: 'fnac',
  manomano: 'manomano',
  amazon: 'amazon',
};

const DELIVERY_RETAILERS = new Set([
  'darty',
  'boulanger',
  'castorama',
  'optimea',
  'amazon',
  'fnac',
  'manomano',
]);

const FETCH_HEADERS = {
  'User-Agent': 'Mozilla/5.0 (iPhone; CPU iPhone OS 17_0 like Mac OS X) AppleWebKit/605.1.15',
  Accept: 'application/json',
  'Accept-Language': 'fr-FR,fr;q=0.9',
};

export interface ClimradarRetailer {
  id: string;
  name: string;
}

export interface ClimradarStore {
  id: string;
  retailerId: string;
  channel: 'online' | 'store';
  name: string;
  country: string;
  address: string;
  city: string;
  postalCode: string;
  department: string;
  lat: number;
  lon: number;
}

export interface ClimradarEntry {
  storeId: string;
  productId: string;
  status: string;
  price: number | null;
  lastSeenMinAgo: number | null;
}

export interface ClimradarStockPayload {
  stores: ClimradarStore[];
  stockByStore: Record<string, ClimradarEntry[]>;
  retailers: ClimradarRetailer[];
  generatedAt?: string;
}

export interface ClimradarLiveRow {
  retailer: ClimradarRetailer;
  store: ClimradarStore;
  entry: ClimradarEntry;
  productUrl: string | null;
}

let stockCache: { fetchedAt: number; payload: ClimradarStockPayload } | null = null;
const STOCK_CACHE_MS = 60_000;

export async function fetchClimradarStock(force = false): Promise<ClimradarStockPayload | null> {
  const now = Date.now();
  if (!force && stockCache && now - stockCache.fetchedAt < STOCK_CACHE_MS) {
    return stockCache.payload;
  }

  try {
    const response = await fetch(CLIMRADAR_STOCK_URL, { headers: FETCH_HEADERS });
    if (!response.ok) return null;
    const payload = (await response.json()) as ClimradarStockPayload;
    if (!payload?.stores?.length || !payload.stockByStore) return null;
    stockCache = { fetchedAt: now, payload };
    return payload;
  } catch {
    return null;
  }
}

export async function fetchClimradarLiveRows(): Promise<ClimradarLiveRow[] | null> {
  try {
    const response = await fetch(CLIMRADAR_LIVE_URL, { headers: FETCH_HEADERS });
    if (!response.ok) return null;
    const json = (await response.json()) as { rows?: ClimradarLiveRow[] };
    return json.rows ?? null;
  } catch {
    return null;
  }
}

function entryForProduct(entries: ClimradarEntry[] | undefined): ClimradarEntry | null {
  if (!entries?.length) return null;
  return entries.find((e) => e.productId === CLIMRADAR_PRODUCT_ID) ?? entries[0];
}

function productUrlMapFromLive(rows: ClimradarLiveRow[] | null): Map<string, string> {
  const map = new Map<string, string>();
  if (!rows) return map;
  for (const row of rows) {
    if (row.productUrl?.startsWith('http')) {
      map.set(row.store.retailerId, row.productUrl);
      map.set(row.store.id, row.productUrl);
    }
  }
  return map;
}

export function isClimradarApiAvailable(payload: ClimradarStockPayload | null): boolean {
  return Boolean(payload?.stores?.length && payload.stockByStore);
}

export function parseOnlineOffersFromStock(
  payload: ClimradarStockPayload,
  liveRows?: ClimradarLiveRow[] | null,
): OnlineOffer[] {
  const storesById = new Map(payload.stores.map((s) => [s.id, s]));
  const retailersById = new Map(payload.retailers.map((r) => [r.id, r]));
  const productUrls = productUrlMapFromLive(liveRows ?? null);
  const offers: OnlineOffer[] = [];
  const seen = new Set<string>();

  for (const [storeId, entries] of Object.entries(payload.stockByStore)) {
    const store = storesById.get(storeId);
    if (!store || store.channel !== 'online' || store.country !== 'FR') continue;

    const entry = entryForProduct(entries);
    if (!entry) continue;

    const retailerId = SLUG_TO_RETAILER_ID[store.retailerId];
    if (!retailerId || seen.has(retailerId)) continue;
    seen.add(retailerId);

    const retailer = retailersById.get(store.retailerId);
    const inStock = entry.status === 'en_stock';
    const climradarUrl = productUrls.get(store.retailerId) ?? productUrls.get(storeId);
    const deliveryEligible =
      DELIVERY_RETAILERS.has(retailerId) && retailerId !== 'leroy_merlin' && inStock;

    offers.push({
      id: retailerId,
      retailerName: retailer?.name ?? store.name.replace(/ en ligne$/i, ''),
      inStock,
      price: entry.price ?? null,
      url: productUrlForRetailer(retailerId, climradarUrl),
      deliveryEligible,
      lastUpdateMin: entry.lastSeenMinAgo ?? null,
    });
  }

  return offers.sort((a, b) => a.retailerName.localeCompare(b.retailerName));
}

export function haversineKm(
  lat1: number,
  lon1: number,
  lat2: number,
  lon2: number,
): number {
  const r = 6371;
  const p1 = (lat1 * Math.PI) / 180;
  const p2 = (lat2 * Math.PI) / 180;
  const dLat = ((lat2 - lat1) * Math.PI) / 180;
  const dLon = ((lon2 - lon1) * Math.PI) / 180;
  const a =
    Math.sin(dLat / 2) ** 2 +
    Math.cos(p1) * Math.cos(p2) * Math.sin(dLon / 2) ** 2;
  return 2 * r * Math.asin(Math.sqrt(a));
}

export function parseLocalStoresFromStock(
  payload: ClimradarStockPayload,
  userLat: number,
  userLon: number,
  radiusKm: number,
  allowedDepts: Set<string>,
  liveRows?: ClimradarLiveRow[] | null,
): LocalStoreOffer[] {
  const storesById = new Map(payload.stores.map((s) => [s.id, s]));
  const productUrls = productUrlMapFromLive(liveRows ?? null);
  const stores: LocalStoreOffer[] = [];
  const seen = new Set<string>();

  for (const [storeId, entries] of Object.entries(payload.stockByStore)) {
    const store = storesById.get(storeId);
    if (!store || store.channel !== 'store' || store.country !== 'FR') continue;

    const retailerId = SLUG_TO_RETAILER_ID[store.retailerId];
    if (!retailerId) continue;

    const entry = entryForProduct(entries);
    if (!entry) continue;

    const dept = store.department || store.postalCode.slice(0, 2);
    if (dept && !allowedDepts.has(dept)) continue;

    const dist = haversineKm(userLat, userLon, store.lat, store.lon);
    if (dist > radiusKm) continue;

    const dedupe = `${retailerId}:${store.postalCode}:${store.name}`;
    if (seen.has(dedupe)) continue;
    seen.add(dedupe);

    const location = store.city
      ? `${store.city.toUpperCase()} ${store.postalCode}`
      : store.postalCode;
    const storePage = store.address.startsWith('http') ? store.address : '';
    const productUrl =
      productUrls.get(storeId) ??
      productUrls.get(store.retailerId) ??
      productUrlForRetailer(retailerId, storePage);
    const safeName = store.name.toLowerCase().replace(/[^a-z0-9]+/g, '-').replace(/^-|-$/g, '');

    stores.push({
      id: `local:${retailerId}:${store.postalCode}:${safeName}`,
      storeName: store.name,
      retailerId,
      location,
      postalCode: store.postalCode,
      department: dept,
      distanceKm: Math.round(dist * 10) / 10,
      inStock: entry.status === 'en_stock',
      price: entry.price ?? null,
      productURL: productUrl,
      lastUpdateMin: entry.lastSeenMinAgo ?? null,
      stockSource: 'climradar',
    });
  }

  return stores.sort((a, b) => {
    if (a.inStock !== b.inStock) return a.inStock ? -1 : 1;
    return a.distanceKm - b.distanceKm;
  });
}

export async function fetchClimradarOnlineOffers(): Promise<{
  offers: OnlineOffer[];
  available: boolean;
}> {
  const [payload, liveRows] = await Promise.all([
    fetchClimradarStock(),
    fetchClimradarLiveRows(),
  ]);
  if (!isClimradarApiAvailable(payload)) {
    return { offers: [], available: false };
  }
  const offers = parseOnlineOffersFromStock(payload!, liveRows);
  return { offers, available: offers.length > 0 };
}
